### PR TITLE
JUnit testing support Bug fixes and Enhancements - Step 1

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR provides a number of bug fixes / enhancements to the current limited JUnit testing support in Che.

More precisely:
- it provides a new ` Tests` text console view to follow in real-time the tests being run
- it fixes problems encountered when trying to run tests on vertx projects:
    -  during the test session the working directory should be set to the project base directory, to be ensure that write access in the current directory is provided to the tested code
    - during the test session the thread context class-loader should be set to the class-loader used to load / run the tested classes (which includes the tested project classpath).
- it reduces the test execution time, by removing the need of running a "maven dependencies" command to build the project test classpath. Instead of that it uses the ` ClasspathService`  to retrieve the already-built classpath.

### What issues does this PR fix or reference?

https://issues.jboss.org/browse/CHE-111
https://issues.jboss.org/browse/CHE-121

#### Changelog
JUnit testing support Bug fixes and Enhancements
- Provided a new ` Tests` text console view.
- Fixed problems encountered when trying to run tests that:
    -  Require write access in the current directory.
    -  Use the thread context class-loader during test execution.
- Reduced the JUnit test execution time by removing an unnecessary step.

#### Release Notes
(Include in release notes only if other PRs are merged in 5.5)

__*Important notice:*__ This PR is the *first step* of series of PR that will all contribute to enhance the testing support. Fro all these steps there will probably be a single release note that would look like this:

Testing support enhancements:
- Real-time notification of test execution,
- Better view of the successful tests,
- Navigation from the test view to the source code,
- various bug fixes

### CODE REVIEW

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml [hunk 1]
```diff
@@ -33,6 +33,10 @@
             <groupId>javax.inject</groupId>
             <artifactId>javax.inject</artifactId>
         </dependency>
+		<dependency>
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml]: sort pom

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml [hunk 2]
```diff
@@ -57,5 +61,23 @@
             <groupId>org.eclipse.che.plugin</groupId>
             <artifactId>che-plugin-testing-classpath-server</artifactId>
         </dependency>
+		<dependency>
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml]: sort pom

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java [hunk 3]
```diff
@@ -148,6 +184,73 @@ private boolean isTestable4x(Class<?> clazz) throws ClassNotFoundException {
...
+	 			    case "testStarted":
+	 				case "testFinished":
+	 				case "testFailure":
+	 				case "testAssumptionFailure":
+	 					return true;
+    	        }
+    	        return false;
+    	    }
+    	});
+		Class<?> c = f.createClass();
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java]: format class

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java [hunk 4]
```diff
@@ -230,19 +350,84 @@ private boolean isTestable3x(Class<?> clazz) throws ClassNotFoundException {
...
+	 			    case "startTest":
+	 				case "endTest":
+	 				case "addError":
+	 				case "addFailure":
+	 					return true;
+    	        }
+    	        return false;
+    	    }
+    	});
+		Class<?> c = f.createClass();
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java]: format class

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/AbstractTestListener.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	public void addFailure(String testKey, Throwable throwable) {
+		getOrCreateTestSummary(testKey).addFailure();
+		addedFailure(testKey, throwable);
+	}
+
+	public void addError(String testKey, Throwable throwable) {
+		getOrCreateTestSummary(testKey).addError();
+		addedError(testKey, throwable);
+	}
+}
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/AbstractTestListener.java]: new line

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		this.stackTraceRoot = strackTraceRoot;
+		writeLine("Starting Test Session", TestingOutput.LineType.SESSION_START);
+	}
+
+	private void writeLine(String line, TestingOutput.LineType lineType) {
+		try {
+			consumer.consume(DtoFactory.cloneDto(new TestingOutputImpl(line, lineType)));
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java]: LOG.error?

File: plugins/plugin-testing/che-plugin-testing-ide/pom.xml [hunk 7]
```diff
@@ -72,6 +72,10 @@
             <groupId>org.eclipse.che.plugin</groupId>
             <artifactId>che-plugin-java-ext-lang-client</artifactId>
         </dependency>
+		<dependency>
+		  <groupId>org.eclipse.che.plugin</groupId>
+		  <artifactId>che-plugin-machine-ext-client</artifactId>
+		</dependency>
```
Comment (skabashnyuk) [inline plugins/plugin-testing/che-plugin-testing-ide/pom.xml]: sort pom

File: wsagent/che-core-api-testing-shared/pom.xml [hunk 8]
```diff
@@ -37,5 +37,17 @@
                 <directory>src/main/java</directory>
             </resource>
         </resources>
+		<plugins>
```
Comment (skabashnyuk) [inline wsagent/che-core-api-testing-shared/pom.xml]: sort pom

File: wsagent/che-core-api-testing/pom.xml [hunk 9]
```diff
@@ -111,6 +111,15 @@
             </resource>
         </resources>
         <plugins>
+		    <plugin>
```
Comment (skabashnyuk) [inline wsagent/che-core-api-testing/pom.xml]: sort pom

**Review by skabashnyuk — CHANGES_REQUESTED** [review]

**Comment by benoitf** [comment]
ci-build

File: plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java [hunk 10]
```diff
@@ -57,35 +65,56 @@ public String getProjectType() {
...
-        classUrls.add(Paths.get(projectPath, "target", "test-classes").toUri().toURL());
-        return classUrls;
+    
+    private Stream<URL> getProjectClasspath(String projectPath) throws JavaModelException {
+    	String relativeProject = projectPath.substring(ResourcesPlugin.getPathToWorkspace().length());
+		IContainer root = ResourcesPlugin.getWorkspace().getRoot();
+    	return toResolvedClassPath(classpathService.getClasspath(relativeProject).stream())
+		.map(dto -> {
+			try {
+				String dtoPath = dto.getPath();
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java]: it seems there is usage of tab characters and indent is not correct

File: wsagent/che-core-api-testing/src/main/java/org/eclipse/che/api/testing/server/handler/TestingOutputImpl.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ * @author David Festal
+ */
+public class TestingOutputImpl implements TestingOutput {
+	
+    public TestingOutputImpl(String output, LineType lineType) {
+		this.output = output;
+		this.lineType = lineType;
+	}
+
+	private String output;
```
Comment (benoitf) [inline wsagent/che-core-api-testing/src/main/java/org/eclipse/che/api/testing/server/handler/TestingOutputImpl.java]: formatting looks odd

**Comment by davidfestal** [comment]
Thanks for these comments @benoitf @skabashnyuk. And sorry for all the formatting problems, that probably mostly come from some bad tab/space eclipse settings. I'll fix all this quickly.

**Comment by benoitf** [comment]
@davidfestal thx

**Comment by davidfestal** [comment]
@benoitf @skabashnyuk I fixed the requested changes.

File: plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java [hunk 12]
```diff
@@ -10,43 +10,51 @@
...
+ * 
  */
 public class MavenTestClasspathProvider implements TestClasspathProvider {
-
-    /**
+	private ClasspathService classpathService;
+	
+	@Inject
+	public MavenTestClasspathProvider(ClasspathService classpathService) {
+		this.classpathService = classpathService;
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java]: @davidfestal I still see here a tab character ?

File: plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java [hunk 13]
```diff
@@ -57,35 +65,56 @@ public String getProjectType() {
...
-        final CommandLine commandLineTestCompile = new CommandLine("mvn", "test-compile");
-        Process processTestCompile = new ProcessBuilder().redirectErrorStream(true).directory(new File(projectPath))
-                .command(commandLineTestCompile.toShellCommand()).start();
-        ProcessUtil.process(processTestCompile, LineConsumer.DEV_NULL, LineConsumer.DEV_NULL);
-        return processTestCompile.waitFor() == 0;
-
+    private Stream<ClasspathEntryDto> toResolvedClassPath(Stream<ClasspathEntryDto> rawClasspath) {
+    	return rawClasspath.flatMap(dto -> {
+    		if (dto.getEntryKind() == IClasspathEntry.CPE_CONTAINER) {
+    			return toResolvedClassPath(dto.getExpandedEntries().stream());
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java]: and there. so I'm not sure it's up-to-date ?

**Comment by davidfestal** [comment]
@benoitf : this time there shouldn't be any tab any more in all the projects I touched !

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunClassContextTestAction.java [hunk 14]
```diff
@@ -46,16 +35,18 @@
...
 
     private final NotificationManager notificationManager;
     private final TestResultPresenter presenter;
-    private final TestServiceClient service;
-    private final AppContext appContext;
-    private final SelectionAgent selectionAgent;
+    private final TestServiceClient   service;
+    private final AppContext          appContext;
+    private final SelectionAgent      selectionAgent;
+    private RunTestActionDelegate     delegate;
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunClassContextTestAction.java]: is there a reason why this delegate is not a final variable ?

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunClassTestAction.java [hunk 15]
```diff
@@ -41,15 +30,17 @@
...
+public class RunClassTestAction extends JavaEditorAction
+                                implements RunTestActionDelegate.Source {
 
     private final NotificationManager notificationManager;
-    private final EditorAgent editorAgent;
+    private final EditorAgent         editorAgent;
     private final TestResultPresenter presenter;
-    private final TestServiceClient service;
+    private final TestServiceClient   service;
+    private RunTestActionDelegate     delegate;
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunClassTestAction.java]: same question there

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunTestActionDelegate.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+        AppContext getAppContext();
+
+        TestServiceClient getService();
+
+        TestResultPresenter getPresenter();
+    }
+
+    public RunTestActionDelegate(
+                                 Source source) {
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-ide/src/main/java/org/eclipse/che/plugin/testing/junit/ide/action/RunTestActionDelegate.java]: it's the formatter that has done this newline break ?

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml [hunk 17]
```diff
@@ -57,5 +61,23 @@
...
+        </dependency>
     </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>com.mycila</groupId>
+                <artifactId>license-maven-plugin</artifactId>
+                <configuration>
+                    <excludes>
+                        <exclude>**/**/AbstractTestListener.java</exclude>
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml:80]: why do we skip these files from the license plugin ?

**Comment by benoitf** [comment]
ci-build

File: plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java [hunk 18]
```diff
@@ -57,35 +64,55 @@ public String getProjectType() {
...
-        String[] paths = line.split(":");
-        for (String path : paths) {
-            classUrls.add(new File(path).toURI().toURL());
-        }
-        bufferedReader.close();
-        fileReader.close();
-        classUrls.add(Paths.get(projectPath, "target", "classes").toUri().toURL());
-        classUrls.add(Paths.get(projectPath, "target", "test-classes").toUri().toURL());
-        return classUrls;
+    private Stream<URL> getProjectClasspath(String projectPath) throws JavaModelException {
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java]: is it possible to have unit tests for that part ?

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Listener for the testing services to report their progress to the Che output view.
+ *
+ * @author David Festal
+ */
+public class OutputTestListener extends AbstractTestListener implements AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractTestListener.class);
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java]: Could we have unit tests ?

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml (hunk 17 above)
Comment (davidfestal) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml:80]: In the new files I created, I added the copyright in the name of *RedHat, Inc*, and the license-checking maven plugin doesn't like that.

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    public OutputTestListener(String strackTraceRoot) {
+        this.stackTraceRoot = strackTraceRoot;
+        writeLine("Starting Test Session", TestingOutput.LineType.SESSION_START);
+    }
+
+    private void writeLine(String line, TestingOutput.LineType lineType) {
+        try {
+            consumer.consume(DtoFactory.cloneDto(new TestingOutputImpl(line, lineType)));
+        } catch (IOException e) {
+            LOG.error("An exception occured while trying to send a 'TestingOutput' object through web sockets on the following channel: " + TESTING_OUTPUT_CHANNEL_NAME, e);
```
Comment (benoitf) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java]: for Log.Error I would prefer to not use string concatenation but instead use of {}
(and well I agree that errors are often written to the log file)

Comment (davidfestal) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java]: fixed

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java (hunk 19 above)
Comment (davidfestal) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java]: That will be, surely, but is it required *now*, or could I push the tests in a next pull request ?

File: plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java (hunk 18 above)
Comment (davidfestal) [reply plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/main/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProvider.java]: That will be, surely, but is it required *now*, or could I push the tests in a next pull request ?

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java (hunk 19 above)
Comment (benoitf) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/OutputTestListener.java]: for me in a next PR is ok

**Review by benoitf — APPROVED** [review]

**Comment by davidfestal** [comment]
@skabashnyuk is it also OK for you ?

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml [hunk 21]
```diff
@@ -57,5 +61,28 @@
...
             <artifactId>che-plugin-testing-classpath-server</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.javassist</groupId>
+            <artifactId>javassist</artifactId>
+        </dependency>
+        <dependency>
+          <groupId>org.slf4j</groupId>
+          <artifactId>slf4j-api</artifactId>
+          <version>1.7.6</version>
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/pom.xml]: remove version.

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/AbstractTestListener.java [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+        getOrCreateTestSummary(testKey).addFailure();
+        addedFailure(testKey, throwable);
+    }
+
+    public void addError(String testKey, Throwable throwable) {
+        getOrCreateTestSummary(testKey).addError();
+        addedError(testKey, throwable);
+    }
+}
+
```
Comment (skabashnyuk) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/listener/AbstractTestListener.java]: I think this extra new line is redundant.

**Review by skabashnyuk — APPROVED** [review]
Ok. after minor cleanups

**Comment by skabashnyuk** [comment]
@bmicklea @JamesDrummond  please review

**Comment by skabashnyuk** [comment]
@davidfestal I would like to wait confirmation from @bmicklea @JamesDrummond about changelog and release notes.

**Comment by davidfestal** [comment]
OK

**Review by l0rd — APPROVED** [review]

**Comment by JamesDrummond** [comment]
Changelog should be a single line. Maybe use the title of PR. 

@skabashnyuk Does this break any of our tests?

**Comment by skabashnyuk** [comment]
I don't know. May be @davidfestal  can explain.

**Review by bmicklea — APPROVED** [review]

**Comment by bmicklea** [comment]
@davidfestal do you expect the other PRs will be ready in the next week?  Or will you need more time?

**Comment by davidfestal** [comment]
The idea is to have most of the other PRs ready at the end of next week yes.

**Comment by TylerJewell** [comment]
@davidfestal - it would be nice to be merge-ready by Thursday COB. We try to close the release every Wednesday, but over the past two releases we have slipped 1-2 days as there are last minute PRs that merge on Monday or Tuesday, and then break the builds, forcing another QA cycle. 

A Thursday night merge gives plenty of time to resolve any outstanding issues in time for the following Wednesday release.

**Review by JamesDrummond — APPROVED** [review]

**Comment by davidfestal** [comment]
I understand the idea @TylerJewell. So in this case, I assume I could simply add my next commits to this PR until Thursday night, so that they can be reviewed as soon as they are added, and we will be ready to merge on Friday ?

**Comment by TylerJewell** [comment]
Yes.  Or if you miss the 5.5 release window, you just merge into the next cycle and it will be part of 5.6 two weeks later.

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java [hunk 23]
```diff
@@ -19,35 +30,24 @@
...
-import static java.util.Collections.emptyList;
-
 /**
  * Service for getting information about classpath.
  *
  * @author Valeriy Svydenko
  */
 @Path("java/classpath/")
-public class ClasspathService {
+public class ClasspathService implements ClasspathServiceInterface {
```
Comment (skabashnyuk) [inline plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java:40]: What is the purpose of interface for rest service with such weird name?

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+        this.apply = apply;
+    }
+    
+    @Override
+    public Return answer(InvocationOnMock invocation) throws Throwable {
+        @SuppressWarnings("unchecked")
+        ArgumentType arg = (ArgumentType)invocation.getArguments()[0];
+        return apply.apply(arg);
+    }
+}
```
Comment (skabashnyuk) [inline plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java]: new line

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+            try {
+                field.setAccessible(true);
+                return field.get(this);
+            } catch (IllegalArgumentException | IllegalAccessException e) {
+                e.printStackTrace();
+                return null;
+            }
+        }).filter(field -> field != null).toArray(Object[]::new));
+    }
+}
```
Comment (skabashnyuk) [inline plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java]: new line

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+import org.mockito.internal.debugging.MockitoDebuggerImpl;
+
+/**
+ * Interface that allows printing the invocations made on
+ * all mocked / spied fields.
+ *
+ * @author David Festal
+ */
+public interface MockitoPrinter {
+    default void printInvocationsOnAllMockedFields() {
```
Comment (skabashnyuk) [inline plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java:24]: this interface look wired. How it will be used?

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+            try {
+                op.apply(error);
+            } catch (OperationException e) {
+                e.printStackTrace();
+            }
+            return promise;
+        }));
+        return this;
+    }
+}
```
Comment (skabashnyuk) [inline plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java]: new line?

**Review by skabashnyuk — COMMENTED** [review]
@davidfestal any plans to finalize this PR so we will be able to made final review?

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java (hunk 23 above)
Comment (davidfestal) [reply plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java:40]: The ` ClasspathService` component is now also used from the server-side class ` org.eclipse.che.plugin.testing.classpath.maven.server.MavenTestClasspathProvider`.

And when implementing the test for `MavenTestClasspathProvider`, mocking  `ClasspathService` directly brought classpath problems, while mocking the implemented interface works perfectly.

see https://github.com/davidfestal/che/blob/5a698ac9aed88f54c6eb98372db493ef39b3ec8b/plugins/plugin-testing-java/plugin-testing-classpath/che-plugin-testing-classpath-maven-server/src/test/java/org/eclipse/che/plugin/testing/classpath/maven/server/MavenTestClasspathProviderTest.java#L46

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java (hunk 26 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java:24]: This interface is used this way:
```Java
public class TestClass implements MockitoPrinter {
  @Mock private aMock1; 
  @Mock private aMock2;
   ...

    @Test
    public void testFunction() {
        ....
        // print all mock invocations called during the test on 
        // mocked fieldsfor debugging and test design purposes  
        printInvocationsOnAllMockedFields();
    }
}
```

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java (hunk 25 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java]: at  the end of the file ?

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java (hunk 27 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java]: at  the end of the file ?

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java (hunk 24 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java]: at the end of the file ?

Comment (skabashnyuk) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java]: yes, Github has nice icon to show that,

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java (hunk 26 above)
Comment (skabashnyuk) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java:24]: Ok. I've missed the fact that is part of the tests

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java (hunk 25 above)
Comment (skabashnyuk) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java]: yes

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java (hunk 27 above)
Comment (skabashnyuk) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java]: yes

**Comment by davidfestal** [comment]
> @davidfestal any plans to finalize this PR so we will be able to made final review?

@skabashnyuk: I plan to further push additional commits very soon, but since this PR is already a consistent whole, if you want I can also push next commits on a separate branch to produce a separate PR.
This way you would all be able to review this PR first.

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java (hunk 27 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/PromiseMocker.java]: new line added

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java (hunk 25 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/MockitoPrinter.java]: new line added

File: plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java (hunk 24 above)
Comment (davidfestal) [reply plugins/plugin-testing/che-plugin-testing-ide/src/test/java/org/eclipse/che/plugin/testing/ide/FunctionAnswer.java]: new line added

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java (hunk 23 above)
Comment (skabashnyuk) [reply plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java:40]: I assume you've added this interface for testing purposes.
From our experience this approche  has some disadvantages.
1. Inheritance in Rest services is quite unusual and we are trying to avoid it in our code.
2. Name of interface is syntetic, it show that it was made for some reason, but not for real inheritance.
3. Mockitng of rest service is quite rare too.

If you do realy want to test Rest service I recomend you to take a look on @Listeners(value = {EverrestJetty.class, MockitoTestNGListener.class})
For example 
https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserServiceTest.java
https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceServiceTest.java

**Review by skabashnyuk — CHANGES_REQUESTED** [review]
I would like to make a review before merge. 
If you are going to add more and more changes I think it would be reasonable to merge Step1 to master, and after create new pr for Step 2 . WDYT?

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java (hunk 23 above)
Comment (davidfestal) [reply plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java:40]: Yes, I added this for testing purposes. The fact is that this rest service was not used by sever-side components until now. It was only used through ` ClasspathServiceClient`  which has not exactly the same method signature (it returns a client promise).

So probably there should be a singleton class, say ` ClasspathProvider`  available for injection in server-side components, and used by the `ClasspathService` rest service.

So do you want me to work on removing ` ClasspathServiceInterface` now in this PR, or can it wait for the next PR ?
In other words does it prevent you to approve the merge of this PR or not ?

**Comment by davidfestal** [comment]
@skabashnyuk I'm OK for merging the PR now and pushing next commits in PR Step 2, sure.
Is there anything you want me to change before merging this PR (see [this comment](https://github.com/eclipse/che/pull/4303#discussion_r105355944))?

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java (hunk 23 above)
Comment (skabashnyuk) [reply plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/ClasspathService.java:40]: Yes, I do want to remove ClasspathServiceInterface. But if you think it'll take significant time you can do that in next PR.

**Review by skabashnyuk — APPROVED** [review]
We good to go.

**Comment by benoitf** [comment]
For merge, please use squash option

**Comment by l0rd** [comment]
@davidfestal build failed please have a look https://ci.codenvycorp.com/job/che-ci-master/1069/

**Comment by davidfestal** [comment]
Yes, it's a problem of undeclared POM transitive dependencies, that make the maven dependency analysis fail.
I already have the fix ready [here](https://github.com/davidfestal/che/commit/f7fab9fb3a0fac50bfeab07f87958eb586c09380) but I'm waiting for the end of a local full build before submitting it in a new PR.

**Comment by davidfestal** [comment]
The PR is submitted: https://github.com/eclipse/che/pull/4391
